File: mifosng-android/src/main/java/com/mifos/mifosxdroid/dialogfragments/syncsurveysdialog/SyncSurveysDialogFragment.java
package com.mifos.mifosxdroid.dialogfragments.syncsurveysdialog;

import android.os.Bundle;
import androidx.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.mifos.mifosxdroid.R;
import com.mifos.mifosxdroid.core.MifosBaseActivity;
import com.mifos.mifosxdroid.core.util.Toaster;
import com.mifos.objects.survey.Survey;
import com.mifos.utils.Constants;
import com.mifos.utils.Network;
import com.mifos.utils.PrefManager;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import androidx.fragment.app.DialogFragment;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;


public class SyncSurveysDialogFragment extends DialogFragment implements SyncSurveysDialogMvpView {

    public static final String LOG_TAG = SyncSurveysDialogFragment.class.getSimpleName();

    @BindView(R.id.tv_sync_title)
    TextView tvSyncTitle;

    @BindView(R.id.tv_survey_name)
    TextView tvSyncingSurveyName;

    @BindView(R.id.tv_total_surveys)
    TextView tvTotalSurveys;

    @BindView(R.id.tv_syncing_survey)
    TextView tvSyncingSurvey;

    @BindView(R.id.pb_sync_survey)
    ProgressBar pbSyncingSurvey;

    @BindView(R.id.tv_total_progress)
    TextView tvTotalProgress;

    @BindView(R.id.pb_total_sync_survey)
    ProgressBar pbTotalSyncSurvey;

    @BindView(R.id.tv_syncing_question)
    TextView tvSyncingQuestion;

    @BindView(R.id.pb_sync_question)
    ProgressBar pbSyncingQuestion;

    @BindView(R.id.tv_syncing_response)
    TextView tvSyncingResponse;

    @BindView(R.id.pb_sync_response)
    ProgressBar pbSyncingResponse;

    @BindView(R.id.tv_sync_failed)
    TextView tvSyncFailed;

    @BindView(R.id.btn_hide)
    Button btnHide;

    @BindView(R.id.btn_cancel)
    Button btnCancel;

    @Inject
    SyncSurveysDialogPresenter mSyncSurveysDialogPresenter;

    View rootView;

    private List<Survey> mSurveyList;


    public static SyncSurveysDialogFragment newInstance() {
        SyncSurveysDialogFragment syncSurveysDialogFragment = new SyncSurveysDialogFragment();
        Bundle args = new Bundle();
        syncSurveysDialogFragment.setArguments(args);
        return syncSurveysDialogFragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        ((MifosBaseActivity) getActivity()).getActivityComponent().inject(this);
        mSurveyList = new ArrayList<Survey>();
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        rootView = inflater.inflate(R.layout.dialog_fragment_sync_surveys, container, false);
        ButterKnife.bind(this, rootView);
        mSyncSurveysDialogPresenter.attachView(this);
        //Start Syncing Surveys
        if (isOnline() && (PrefManager.getUserStatus() == Constants.USER_ONLINE)) {
            mSyncSurveysDialogPresenter.loadSurveyList();
        } else {
            showNetworkIsNotAvailable();
            getFragmentManager().popBackStack();
        }

        return rootView;
    }

    @OnClick(R.id.btn_cancel)
    void onClickCancelButton() {
        dismissDialog();
    }

    @OnClick(R.id.btn_hide)
    void onClickHideButton() {
        if (btnHide.getText().equals(getResources().getString(R.string.dialog_action_ok))) {
            dismissDialog();
        } else {
            hideDialog();
        }
    }

    @Override
    public void updateSurveyList(List<Survey> surveyList) {
        mSurveyList = surveyList;
    }

    @Override
    public void showUI() {
        pbTotalSyncSurvey.setMax(mSurveyList.size());
        String total_surveys = mSurveyList.size() + getResources().getString(R.string.space) +
                getResources().getString(R.string.surveys);
        tvTotalSurveys.setText(total_surveys);
        tvSyncFailed.setText(String.valueOf(0));
    }

    @Override
    public void showSyncingSurvey(String surveyName) {
        tvSyncingSurvey.setText(surveyName);
        tvSyncingSurveyName.setText(surveyName);
    }

    @Override
    public void showSyncedFailedSurveys(int failedCount) {
        tvSyncFailed.setText(String.valueOf(failedCount));
    }

    @Override
    public void setMaxSingleSyncSurveyProgressBar(int total) {
        pbSyncingSurvey.setMax(total);
    }

    @Override
    public void setQuestionSyncProgressBarMax(int count) {
        pbSyncingQuestion.setMax(count);
    }

    @Override
    public void setResponseSyncProgressBarMax(int count) {
        pbSyncingResponse.setMax(count);
    }

    @Override
    public void updateSingleSyncSurveyProgressBar(int count) {
        pbSyncingSurvey.setProgress(count);
    }

    @Override
    public void updateQuestionSyncProgressBar(int i) {
        pbSyncingQuestion.setProgress(i);
    }

    @Override
    public void updateResponseSyncProgressBar(int i) {
        pbSyncingResponse.setProgress(i);
    }

    @Override
    public void updateTotalSyncSurveyProgressBarAndCount(int count) {
        pbTotalSyncSurvey.setProgress(count);
        String total_sync_count = getResources()
                .getString(R.string.space) + count + getResources()
                .getString(R.string.slash) + mSurveyList.size();
        tvTotalProgress.setText(total_sync_count);
    }

    @Override
    public int getMaxSingleSyncSurveyProgressBar() {
        return pbSyncingSurvey.getMax();
    }

    @Override
    public void showNetworkIsNotAvailable() {
        Toast.makeText(getActivity(), getResources().getString(R.string
                .error_network_not_available), Toast.LENGTH_SHORT).show();
    }

    @Override
    public void showSurveysSyncSuccessfully() {
        btnCancel.setVisibility(View.INVISIBLE);
        dismissDialog();
        Toast.makeText(getActivity(), R.string.sync_success, Toast.LENGTH_SHORT).show();
    }

    @Override
    public Boolean isOnline() {
        return Network.isOnline(getActivity());
    }

    @Override
    public void dismissDialog() {
        getDialog().dismiss();
    }

    @Override
    public void showDialog() {
        getDialog().show();
    }

    @Override
    public void hideDialog() {
        getDialog().hide();
    }

    @Override
    public void showError(int s) {
        Toaster.show(rootView, s);
    }

    @Override
    public void showProgressbar(boolean b) {

    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mSyncSurveysDialogPresenter.detachView();
    }
}

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/dialogfragments/syncsurveysdialog/SyncSurveysDialogFragment.kt
package com.mifos.mifosxdroid.dialogfragments.syncsurveysdialog

/*
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.DialogFragment
import androidx.lifecycle.ViewModelProvider
import com.mifos.core.objects.survey.Survey
import com.mifos.mifosxdroid.R
import com.mifos.mifosxdroid.core.util.Toaster.show
import com.mifos.mifosxdroid.databinding.DialogFragmentSyncSurveysBinding
import com.mifos.utils.Constants
import com.mifos.utils.Network.isOnline
import com.mifos.utils.PrefManager.userStatus
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class SyncSurveysDialogFragment : DialogFragment() {

    private lateinit var binding: DialogFragmentSyncSurveysBinding

    private lateinit var viewModel: SyncSurveysDialogViewModel

    private var mSurveyList: List<Survey>? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        mSurveyList = ArrayList()
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = DialogFragmentSyncSurveysBinding.inflate(inflater, container, false)
        viewModel = ViewModelProvider(this)[SyncSurveysDialogViewModel::class.java]
        //Start Syncing Surveys
        if (isOnline && userStatus == Constants.USER_ONLINE) {
            viewModel.loadSurveyList()
        } else {
            showNetworkIsNotAvailable()
            fragmentManager?.popBackStack()
        }

        viewModel.syncSurveysDialogUiState.observe(viewLifecycleOwner) {
            when (it) {
                is SyncSurveysDialogUiState.DismissDialog -> hideDialog()
                is SyncSurveysDialogUiState.SetQuestionSyncProgressBarMax -> {
                    showProgressbar(false)
                    setQuestionSyncProgressBarMax(it.total)
                }

                is SyncSurveysDialogUiState.SetResponseSyncProgressBarMax -> {
                    showProgressbar(false)
                    setResponseSyncProgressBarMax(it.total)
                }

                is SyncSurveysDialogUiState.ShowError -> {
                    showProgressbar(false)
                    showError(it.message)
                }

                is SyncSurveysDialogUiState.ShowNetworkIsNotAvailable -> {
                    showProgressbar(false)
                    showNetworkIsNotAvailable()
                }

                is SyncSurveysDialogUiState.ShowProgressbar -> showProgressbar(true)
                is SyncSurveysDialogUiState.ShowSurveysSyncSuccessfully -> {
                    showProgressbar(false)
                    showSurveysSyncSuccessfully()
                }

                is SyncSurveysDialogUiState.ShowSyncedFailedSurveys -> {
                    showProgressbar(false)
                    showSyncedFailedSurveys(it.failedCount)
                }

                is SyncSurveysDialogUiState.ShowSyncingSurvey -> {
                    showProgressbar(false)
                    showSyncingSurvey(it.surveyName)
                }

                is SyncSurveysDialogUiState.ShowUI -> {
                    showProgressbar(false)
                    showUI()
                }

                is SyncSurveysDialogUiState.UpdateQuestionSyncProgressBar -> {
                    showProgressbar(false)
                    updateQuestionSyncProgressBar(it.index)
                }

                is SyncSurveysDialogUiState.UpdateResponseSyncProgressBar -> {
                    showProgressbar(false)
                    updateResponseSyncProgressBar(it.index)
                }

                is SyncSurveysDialogUiState.UpdateSingleSyncSurveyProgressBar -> {
                    showProgressbar(false)
                    updateSingleSyncSurveyProgressBar(it.index)
                }

                is SyncSurveysDialogUiState.UpdateSurveyList -> {
                    showProgressbar(false)
                    updateSurveyList(it.surveys)
                }

                is SyncSurveysDialogUiState.UpdateTotalSyncSurveyProgressBarAndCount -> {
                    showProgressbar(false)
                    updateTotalSyncSurveyProgressBarAndCount(it.total)
                }

                SyncSurveysDialogUiState.Initial -> Unit
            }
        }

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.btnCancel.setOnClickListener {
            onClickCancelButton()
        }

        binding.btnHide.setOnClickListener {
            onClickHideButton()
        }
    }

    private fun onClickCancelButton() {
        dismissDialog()
    }

    private fun onClickHideButton() {
        if (binding.btnHide.text == resources.getString(R.string.dialog_action_ok)) {
            dismissDialog()
        } else {
            hideDialog()
        }
    }

    private fun updateSurveyList(surveyList: List<Survey>) {
        mSurveyList = surveyList
    }

    private fun showUI() {
        binding.pbTotalSyncSurvey.max = mSurveyList!!.size
        val totalSurveys = mSurveyList!!.size.toString() + resources.getString(R.string.space) +
                resources.getString(R.string.surveys)
        binding.tvTotalSurveys.text = totalSurveys
        binding.tvSyncFailed.text = 0.toString()
    }

    private fun showSyncingSurvey(surveyName: String?) {
        binding.tvSyncingSurvey.text = surveyName
        binding.tvSurveyName.text = surveyName
    }

    private fun showSyncedFailedSurveys(failedCount: Int) {
        binding.tvSyncFailed.text = failedCount.toString()
    }

    private fun setQuestionSyncProgressBarMax(count: Int) {
        binding.pbSyncQuestion.max = count
    }

    private fun setResponseSyncProgressBarMax(count: Int) {
        binding.pbSyncResponse.max = count
    }

    private fun updateSingleSyncSurveyProgressBar(count: Int) {
        binding.pbSyncSurvey.progress = count
    }

    private fun updateQuestionSyncProgressBar(i: Int) {
        binding.pbSyncQuestion.progress = i
    }

    private fun updateResponseSyncProgressBar(i: Int) {
        binding.pbSyncResponse.progress = i
    }

    private fun updateTotalSyncSurveyProgressBarAndCount(count: Int) {
        binding.pbTotalSyncSurvey.progress = count
        val totalSyncCount = resources
            .getString(R.string.space) + count + resources
            .getString(R.string.slash) + mSurveyList!!.size
        binding.tvTotalProgress.text = totalSyncCount
    }

    private var maxSingleSyncSurveyProgressBar: Int
        get() = binding.pbSyncSurvey.max
        set(total) {
            binding.pbSyncSurvey.max = total
        }

    private fun showNetworkIsNotAvailable() {
        Toast.makeText(
            activity,
            resources.getString(R.string.error_network_not_available),
            Toast.LENGTH_SHORT
        ).show()
    }

    private fun showSurveysSyncSuccessfully() {
        binding.btnCancel.visibility = View.INVISIBLE
        dismissDialog()
        Toast.makeText(activity, R.string.sync_success, Toast.LENGTH_SHORT).show()
    }

    private val isOnline: Boolean
        get() = activity?.let { isOnline(it) } == true

    private fun dismissDialog() {
        dialog?.dismiss()
    }

    private fun showDialog() {
        dialog?.show()
    }

    private fun hideDialog() {
        dialog?.hide()
    }

    private fun showError(s: String) {
        show(binding.root, s)
    }

    private fun showProgressbar(b: Boolean) {}

    companion object {
        val LOG_TAG = SyncSurveysDialogFragment::class.java.simpleName
        fun newInstance(): SyncSurveysDialogFragment {
            val syncSurveysDialogFragment = SyncSurveysDialogFragment()
            val args = Bundle()
            syncSurveysDialogFragment.arguments = args
            return syncSurveysDialogFragment
        }
    }
}
 */